### ISSUE TITLE
Updating states from randomly generated objects and adding new associations

### ISSUE THREAD

**Issue by eagleEggs** [Unity-Technologies/ml-agents#126]
Could you assist briefly with a summary of how you would approach this type of scenario, thanks:

1. There are only two states initially, success and failure, which are consistently randomly altered floats.
2. There are objects of many different randomized types which do not have states, but need to become states when a failure occurs (reward = -1). i.e. associating the failure with the completely new and random variable caused the failure. That randomized object then becomes a state and if encountered again, the AI should learn to stay away from that objects float value.

How would you go about dynamically aggregating states, from randomized objects that could be anything that caused the failure float to trigger?

If it's unclear I'll try to clean up my example here :D thanks again

**Comment by eagleEggs** [comment]
I guess a way to visualize would be that during a game, you would instantiate a completely new prefab that has no state float value, and want to add that to a new state that would be learned over time.

**Comment by eagleEggs** [comment]
Actually I believe I figured it out. There would just be empty states, until updated from an outside class. However is there any way to call agent.state.Add and recreate the list at runtime?

**Comment by surferau** [comment]
Negative reward its a trap a got into as well. People who does know a little about neural networks in general and Q reinforcement learning in particular may understand this wrong.

First of all, everything is relative. -1 and 1 reward is about equal as 0 and 2 reward(in terms of maximum value).
Second, everything is about total reward(Cumulative reward). The agent tries always choose a path of statements->decision with maximal total reward.

If you have  a lot of states it might be better to use Discrete state representation (one integer describes all states and its your job how to encode them) .

>However is there any way to call agent.state.Add and recreate the list at runtime?

I am not sure what you asking for. After CollectStates is called, the collected states are being sent to python code where network decides what action to take and sends the action back. So when are in AgentStep, states were already used. If you need to keep copy of them you can add a property to agent class and do something like this in "CollectStates"
List<float> agentStates = new List<float>(states) or
states = new List<float>(generatedStates);
Just remember that states can be modified only during CollectStates call after that they are discarded.

I must tell after a lot of experimentation with mlagents, it is not yet ready to use in production. It takes ungodly amount of time to train anything and results kinda meh. Alternative is to use TensorFlowSharp plugin+mlagents inside unity, but generate the .bytes file with simple network and not reinforced learning time.

**Comment by eagleEggs** [comment]
Thanks for the info @surferau !!! So let me illustrate a bit more about what I'm trying to do and see if you or anyone else has any insight. I think I rushed my comment and with what you said I may not be able to do it the way I was thinking:

1. Within the game I'm streaming floats from an XML into an array and iterating through it to apply a transform.position to a gameObject. That gameObjects X and Y would be a state in UML.

2. There are two actions for that gameObject: start and stop.

3. There is another layer of data from another XML iterating in an array, which represents the phases of the moon over time as a float. That would also be a state.

4. When the agent decides to start or stop, the learned behavior is to identify when the moon is affecting the transform.position of the gameObject.

That's what I already understand how to do.

However what about this:

1. Within the app after build (within the code), I will be streaming new data sets from XML into arrays. This data will be seemingly random and have no association with the gameObject nor the moon data. However I'd want the agent to treat that new data the same way as the moon data, and possibly find associations itself. So, basically, streaming new states into the agent while maintaining what it has learned from the previous states and just applying the new data sets going forward.

Still not sure if I'm explaining this enough lol.... thanks!

**Comment by surferau** [comment]
A bit hard to understand but lets try. SInce agentstep is called multiple times during a frame (because its being called from FixedUpdate) you can split your decision making process into phases which you control with some kind of global state: phase1 collect this states, get action1; phase2 collect other states get action2; last phase reset and start from the beginning.

Or

You can have several branes with different states set and work independently and generate set of different actions (one brain responsible for moon phase, other for something else)

The library is very early stage of development and not everything is possible. I pretty much failed myself to get any acceptable results with it. While it took a couple of settings to train simple network on generated set of data, i was not able to train PPO model with same set of data at all.

**Comment by eagleEggs** [comment]
Thanks for the info @surferau !!!
